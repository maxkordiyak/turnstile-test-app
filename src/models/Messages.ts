export enum MESSAGES {
  THANK_YOU = "Thank you for using our Metro services. You can now pass the turnstile",
  LOCKED = "Please insert coins to pass",
  ERROR = "The machine has been broken. Please fix it"
}
