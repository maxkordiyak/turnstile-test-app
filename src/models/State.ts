export interface State {
    locked: boolean,
    message: string,
    amount: number,
    spinning: boolean,
    error: boolean
}