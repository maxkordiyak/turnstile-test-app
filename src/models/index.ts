export * from './State';
export * from './Messages';