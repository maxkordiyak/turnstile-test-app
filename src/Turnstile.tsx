import React, { Component } from "react";
import { Spinning } from "./components/Spinning";
import { State, MESSAGES } from "./models";

export default class Turnstile extends Component<any, State> {
  private readonly correctAmount: number;

  constructor(props: any) {
    super(props);

    this.state = {
      locked: true,
      message: MESSAGES.LOCKED,
      amount: 0,
      spinning: false,
      error: false
    };

    this.correctAmount = 3;
    this.insertCoin = this.insertCoin.bind(this);
    this.pass = this.pass.bind(this);
    this.break = this.break.bind(this);
    this.fix = this.fix.bind(this);
  }

  componentDidUpdate(): void {
    if (this.state.locked && this.state.amount >= this.correctAmount) {
      this.setState({ locked: false, message: MESSAGES.THANK_YOU });
    }
  }

  insertCoin(): void {
    this.setState(prevState => {
      return {
        amount: prevState.amount + 1
      };
    });
  }

  pass(): void {
    this.setState(
      {
        locked: true,
        amount: 0,
        spinning: true
      },
      () => {
        setTimeout(() => {
          this.setState({
            message: MESSAGES.LOCKED,
            spinning: false
          });
        }, 2000);
      }
    );
  }

  break(): void {
    this.setState({ error: true });
  }

  fix(): void {
    this.setState({
      error: false
    });
  }

  render() {
    const { message, amount, locked, spinning, error } = this.state;

    let errorMessage: object | null = null;
    if (error) {
      errorMessage = (
        <div className="error">
          <p>{MESSAGES.ERROR}</p>
          <button onClick={this.fix}>Fix</button>
        </div>
      );
    }

    return (
      <div className="app">
        <div>
          <h1>Turnstile state machine</h1>
          <h2>Subway fare: ${this.correctAmount}</h2>
          <p>{message}</p>
          <p>Amount of coins inserted: ${amount}</p>
          <div className="actions">
            <p>
              <button
                disabled={error}
                onClick={spinning ? this.break : this.insertCoin}
              >
                Insert a coin
              </button>
            </p>
            <p>
              <button disabled={locked || error} onClick={this.pass}>
                Walk - pass the turnstile
              </button>
            </p>
          </div>
          {spinning && <Spinning />}
          {errorMessage}
          <small>{JSON.stringify(this.state)}</small>
        </div>
      </div>
    );
  }
}
