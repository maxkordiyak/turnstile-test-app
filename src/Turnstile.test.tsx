import React from "react";
import Turnstile from "./Turnstile";
import { shallow } from "enzyme";
import { MESSAGES } from "./models";

jest.useFakeTimers();

it("renders Turnstile properly", () => {
  const component = getComponent();

  expect(component).toMatchSnapshot();
});

it("should increase amount when user presses `Insert a coin` button", () => {
  const component = getComponent();

  const insertButton = component.find("button").at(0);
  insertButton.simulate("click");

  expect(component.find("p").at(1).text()).toEqual("Amount of coins inserted: $1");

  insertButton.simulate("click");
  expect(component.find("p").at(1).text()).toEqual("Amount of coins inserted: $2");
});

it("should display an error when user tried to insert a coin while passing", () => {
  const component = getComponent();
  expect(component.exists(".error")).toEqual(false);

  component.find("button").at(0).simulate("click");
  component.find("button").at(0).simulate("click");
  component.find("button").at(0).simulate("click");

  component.find("button").at(1).simulate("click");
  component.find("button").at(0).simulate("click");

  expect(component.find("button").at(0).prop("disabled")).toBe(true);
  expect(component.find("button").at(1).prop("disabled")).toBe(true);

  expect(component.exists(".error")).toEqual(true);
});

it("hides an error when `Fix` button is pressed", () => {
  const component = getComponent();
  expect(component.exists(".error")).toEqual(false);

  component.find("button").at(0).simulate("click");
  component.find("button").at(0).simulate("click");
  component.find("button").at(0).simulate("click");

  component.find("button").at(1).simulate("click");
  component.find("button").at(0).simulate("click");

  expect(component.exists(".error")).toEqual(true);
  component.find("button").at(2).simulate("click");
  expect(component.exists(".error")).toEqual(false);
});

it("shows message when user passes", () => {
  const component = getComponent();

  component.find("button").at(0).simulate("click");
  component.find("button").at(0).simulate("click");
  component.find("button").at(0).simulate("click");
  component.find("button").at(1).simulate("click");
  expect(component.find("p").at(0).text()).toEqual(MESSAGES.THANK_YOU);
});


const getComponent = () => {
  return shallow(<Turnstile />);
};
