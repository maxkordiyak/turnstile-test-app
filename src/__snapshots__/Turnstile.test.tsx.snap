// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Turnstile properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Turnstile />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <h1>
          Turnstile state machine
        </h1>
        <h2>
          Subway fare: $
          3
        </h2>
        <p>
          Please insert coins to pass
        </p>
        <p>
          Amount of coins inserted: $
          0
        </p>
        <div
          className="actions"
        >
          <p>
            <button
              disabled={false}
              onClick={[Function]}
            >
              Insert a coin
            </button>
          </p>
          <p>
            <button
              disabled={true}
              onClick={[Function]}
            >
              Walk - pass the turnstile
            </button>
          </p>
        </div>
        <small>
          {"locked":true,"message":"Please insert coins to pass","amount":0,"spinning":false,"error":false}
        </small>
      </div>,
      "className": "app",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Turnstile state machine
          </h1>,
          <h2>
            Subway fare: $
            3
          </h2>,
          <p>
            Please insert coins to pass
          </p>,
          <p>
            Amount of coins inserted: $
            0
          </p>,
          <div
            className="actions"
          >
            <p>
              <button
                disabled={false}
                onClick={[Function]}
              >
                Insert a coin
              </button>
            </p>
            <p>
              <button
                disabled={true}
                onClick={[Function]}
              >
                Walk - pass the turnstile
              </button>
            </p>
          </div>,
          false,
          null,
          <small>
            {"locked":true,"message":"Please insert coins to pass","amount":0,"spinning":false,"error":false}
          </small>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Turnstile state machine",
          },
          "ref": null,
          "rendered": "Turnstile state machine",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Subway fare: $",
              3,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Subway fare: $",
            3,
          ],
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Please insert coins to pass",
          },
          "ref": null,
          "rendered": "Please insert coins to pass",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Amount of coins inserted: $",
              0,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Amount of coins inserted: $",
            0,
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                <button
                  disabled={false}
                  onClick={[Function]}
                >
                  Insert a coin
                </button>
              </p>,
              <p>
                <button
                  disabled={true}
                  onClick={[Function]}
                >
                  Walk - pass the turnstile
                </button>
              </p>,
            ],
            "className": "actions",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <button
                  disabled={false}
                  onClick={[Function]}
                >
                  Insert a coin
                </button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Insert a coin",
                  "disabled": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Insert a coin",
                "type": "button",
              },
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <button
                  disabled={true}
                  onClick={[Function]}
                >
                  Walk - pass the turnstile
                </button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Walk - pass the turnstile",
                  "disabled": true,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Walk - pass the turnstile",
                "type": "button",
              },
              "type": "p",
            },
          ],
          "type": "div",
        },
        false,
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "{\\"locked\\":true,\\"message\\":\\"Please insert coins to pass\\",\\"amount\\":0,\\"spinning\\":false,\\"error\\":false}",
          },
          "ref": null,
          "rendered": "{\\"locked\\":true,\\"message\\":\\"Please insert coins to pass\\",\\"amount\\":0,\\"spinning\\":false,\\"error\\":false}",
          "type": "small",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <h1>
            Turnstile state machine
          </h1>
          <h2>
            Subway fare: $
            3
          </h2>
          <p>
            Please insert coins to pass
          </p>
          <p>
            Amount of coins inserted: $
            0
          </p>
          <div
            className="actions"
          >
            <p>
              <button
                disabled={false}
                onClick={[Function]}
              >
                Insert a coin
              </button>
            </p>
            <p>
              <button
                disabled={true}
                onClick={[Function]}
              >
                Walk - pass the turnstile
              </button>
            </p>
          </div>
          <small>
            {"locked":true,"message":"Please insert coins to pass","amount":0,"spinning":false,"error":false}
          </small>
        </div>,
        "className": "app",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Turnstile state machine
            </h1>,
            <h2>
              Subway fare: $
              3
            </h2>,
            <p>
              Please insert coins to pass
            </p>,
            <p>
              Amount of coins inserted: $
              0
            </p>,
            <div
              className="actions"
            >
              <p>
                <button
                  disabled={false}
                  onClick={[Function]}
                >
                  Insert a coin
                </button>
              </p>
              <p>
                <button
                  disabled={true}
                  onClick={[Function]}
                >
                  Walk - pass the turnstile
                </button>
              </p>
            </div>,
            false,
            null,
            <small>
              {"locked":true,"message":"Please insert coins to pass","amount":0,"spinning":false,"error":false}
            </small>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Turnstile state machine",
            },
            "ref": null,
            "rendered": "Turnstile state machine",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Subway fare: $",
                3,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Subway fare: $",
              3,
            ],
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Please insert coins to pass",
            },
            "ref": null,
            "rendered": "Please insert coins to pass",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Amount of coins inserted: $",
                0,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Amount of coins inserted: $",
              0,
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p>
                  <button
                    disabled={false}
                    onClick={[Function]}
                  >
                    Insert a coin
                  </button>
                </p>,
                <p>
                  <button
                    disabled={true}
                    onClick={[Function]}
                  >
                    Walk - pass the turnstile
                  </button>
                </p>,
              ],
              "className": "actions",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <button
                    disabled={false}
                    onClick={[Function]}
                  >
                    Insert a coin
                  </button>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Insert a coin",
                    "disabled": false,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Insert a coin",
                  "type": "button",
                },
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <button
                    disabled={true}
                    onClick={[Function]}
                  >
                    Walk - pass the turnstile
                  </button>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Walk - pass the turnstile",
                    "disabled": true,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Walk - pass the turnstile",
                  "type": "button",
                },
                "type": "p",
              },
            ],
            "type": "div",
          },
          false,
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "{\\"locked\\":true,\\"message\\":\\"Please insert coins to pass\\",\\"amount\\":0,\\"spinning\\":false,\\"error\\":false}",
            },
            "ref": null,
            "rendered": "{\\"locked\\":true,\\"message\\":\\"Please insert coins to pass\\",\\"amount\\":0,\\"spinning\\":false,\\"error\\":false}",
            "type": "small",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
